pyteal/ast/abi/array.py

```diff
@@ -0,0 +1,217 @@
+            ):
+                raise TealInputError("Index out of bounds")
+            index = Int(index)
+        return ArrayElement(self, cast(Expr, index), length)
+
+    def __str__(self) -> str:
+        return self._valueType.__str__() + (
+            "[]" if self._static_length is None else "[{}]".format(self._static_length)
+        )
+
+
+Array.__module__ = "pyteal"
+
+# until something like https://github.com/python/mypy/issues/3345 is added, we can't make the size of the array a generic parameter
+class StaticArray(Array[T]):
+    def __init__(self, valueType: T, length: int) -> None:
+        super().__init__(valueType, length)
+
+    def has_same_type_as(self, other: Type) -> bool:
+        return (
+            type(other) is StaticArray
+            and self._valueType.has_same_type_as(other._valueType)
+            and self.length_static() == other.length_static()
+        )
+
+    def new_instance(self) -> "StaticArray[T]":
+        return StaticArray(self._valueType, self.length_static())
+
+    def is_dynamic(self) -> bool:
+        return self._valueType.is_dynamic()
+
+    def byte_length_static(self) -> int:
+        if self.is_dynamic():
+            raise ValueError("Type is dynamic")
+        if type(self._valueType) is Bool:
+            return boolSequenceLength(self.length_static())
+        return self.length_static() * self._valueType.byte_length_static()
+
+    def set(self, values: Sequence[T]) -> Expr:
+        if self.length_static() != len(values):
+            raise TealInputError(
+                "Incorrect length for values. Expected {}, got {}".format(
+                    self.length_static(), len(values)
+                )
+            )
+        return super().set(values)
+
+    def length_static(self) -> int:
+        return cast(int, self._static_length)
+
+    def length(self) -> Expr:
+        return Int(self.length_static())
+
+
+StaticArray.__module__ = "pyteal"
+
+
+class DynamicArray(Array[T]):
+    def __init__(self, valueType: T) -> None:
+        super().__init__(valueType, None)
+
+    def has_same_type_as(self, other: Type) -> bool:
+        return type(other) is DynamicArray and self._valueType.has_same_type_as(
+            other._valueType
+        )
+
+    def new_instance(self) -> "DynamicArray[T]":
+        return DynamicArray(self._valueType)
+
+    def is_dynamic(self) -> bool:
+        return True
+
+    def byte_length_static(self) -> int:
+        raise ValueError("Type is dynamic")
+
+    def length(self) -> Expr:
+        output = Uint16()
+        return Seq(
+            output.decode(self.encode()),
+            output.get(),
+        )
+
+
+DynamicArray.__module__ = "pyteal"
+
+
+class ArrayElement(ComputedType[T]):
+    def __init__(self, array: Array[T], index: Expr, length: Expr = None) -> None:
+        super().__init__(array._valueType)
+        self.array = array
+        self.index = index
+        self.length = length
```

Comment:
Good point, this is leftover when when we had `Array.map` and `Array.forEach`. In those cases, it's more efficient to calculate the length once and then reuse it every time you need to load an element. But since we don't have those functions anymore, I will remove this